### PR TITLE
bpo-32077: Documentation: Some Unicode object functions don't indicate whether they return a new reference

### PR DESCRIPTION
**Pull request by lioncash**
Also adds missing functions. Makes the documentation more comprehensive in terms of indicating whether or not a function returns a new reference.

I've gone over the entries quite a few times, and I'm fairly confident I've amended most of the existing functions and added the new ones that weren't yet added. If I've happened to miss any, just say the word and I'll add them.

I've also noticed the API documentation for bytes objects is also missing entries, which I've made a local commit for, but wasn't sure if it was a good idea to intermix it in this change. I can PR it separately or include it in this one depending on whichever is preferable.

https://bugs.python.org/issue32077

### CODE REVIEW

File: Doc/data/refcounts.dat [hunk 1]
```diff
@@ -1503,35 +1514,39 @@ PyUnicode_FromEncodedObject:const char*:errors::
 
 PyUnicode_FromWideChar:PyObject*::+1:
 PyUnicode_FromWideChar:const wchar_t*:w::
-PyUnicode_FromWideChar:int:size::
+PyUnicode_FromWideChar:Py_ssize_t:size::
 
-PyUnicode_AsWideChar:int:::
-PyUnicode_AsWideChar:PyObject*:*unicode:0:
+PyUnicode_AsWideChar:Py_ssize_t:::
+PyUnicode_AsWideChar:PyUnicodeObject*:unicode:0:
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1520]: `PyObject`

File: Doc/data/refcounts.dat [hunk 2]
```diff
@@ -1503,35 +1514,39 @@ PyUnicode_FromEncodedObject:const char*:errors::
...
 PyUnicode_Decode:const char*:encoding::
 PyUnicode_Decode:const char*:errors::
 
 PyUnicode_DecodeUTF16Stateful:PyObject*::+1:
 PyUnicode_DecodeUTF16Stateful:const char*:s::
-PyUnicode_DecodeUTF16Stateful:int:size::
+PyUnicode_DecodeUTF16Stateful:Py_ssize_t:size::
 PyUnicode_DecodeUTF16Stateful:const char*:errors::
 PyUnicode_DecodeUTF16Stateful:int*:byteorder::
 PyUnicode_DecodeUTF16Stateful:int*:consumed::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1539]: `Py_ssize_t`

File: Doc/data/refcounts.dat [hunk 3]
```diff
@@ -1503,35 +1514,39 @@ PyUnicode_FromEncodedObject:const char*:errors::
...
 
 PyUnicode_DecodeUTF8Stateful:PyObject*::+1:
 PyUnicode_DecodeUTF8Stateful:const char*:s::
-PyUnicode_DecodeUTF8Stateful:int:size::
+PyUnicode_DecodeUTF8Stateful:Py_ssize_t:size::
 PyUnicode_DecodeUTF8Stateful:const char*:errors::
 PyUnicode_DecodeUTF8Stateful:int*:consumed::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1545]: `Py_ssize_t`

File: Doc/data/refcounts.dat [hunk 4]
```diff
@@ -1540,95 +1555,140 @@ PyUnicode_AsEncodedString:PyObject*:unicode::
...
+PyUnicode_EncodeUTF8:Py_ssize_t:size::
 PyUnicode_EncodeUTF8:const char*:errors::
 
 PyUnicode_AsUTF8String:PyObject*::+1:
-PyUnicode_AsUTF8String:PyObject*:unicode::
+PyUnicode_AsUTF8String:PyObject*:unicode:0:
+
+PyUnicode_AsUTF8AndSize:const char*:::
+PyUnicode_AsUTF8AndSize:PyObject*:unicode:0:
+
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1591]: `PyUnicode_AsUTF8AndSize:Py_ssize_t*:size::`

File: Doc/data/refcounts.dat [hunk 5]
```diff
@@ -1540,95 +1555,140 @@ PyUnicode_AsEncodedString:PyObject*:unicode::
...
 PyUnicode_DecodeRawUnicodeEscape:const char*:s::
-PyUnicode_DecodeRawUnicodeEscape:int:size::
+PyUnicode_DecodeRawUnicodeEscape:Py_ssize_t:size::
 PyUnicode_DecodeRawUnicodeEscape:const char*:errors::
 
 PyUnicode_EncodeRawUnicodeEscape:PyObject*::+1:
 PyUnicode_EncodeRawUnicodeEscape:const Py_UNICODE*:s::
-PyUnicode_EncodeRawUnicodeEscape:int:size::
+PyUnicode_EncodeRawUnicodeEscape:Py_ssize_t:size::
 PyUnicode_EncodeRawUnicodeEscape:const char*:errors::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1652]: Remove.

File: Doc/data/refcounts.dat [hunk 6]
```diff
@@ -1715,6 +1802,105 @@ PyUnicode_Contains:int:::
...
 PyUnicode_Contains:PyObject*:element:0:
 
+PyUnicode_InternInPlace:void:::
+PyUnicode_InternInPlace:PyObject**:string:+1:
+
+PyUnicode_InternFromString:PyObject*::+1:
+PyUnicode_InternFromString:const char*:cp::
+
+PyUnicode_New:PyObject*::+1:
+PyUnicode_New:Py_UCS4:maxchar::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1812]: Insert `PyUnicode_New:Py_ssize_t:size::`.

**Comment by lioncash** [comment]
I've recreated the pull request in #11243, with the feedback addressed, as I mistakenly thought this was addressed in the past and removed my fork on github, which seems to have orphaned this pull request. I'm really sorry about that.

**Pull request closed without merging**